### PR TITLE
[PFC] Add a pytest test for PFC counters

### PR DESCRIPTION
**Pull request by baiwei0427**
### Description of PR

Summary: A pytest test to validate SONiC PFC counters

### Type of change
- [] Test case(new/improvement)

#### How did you do it?
- Added sonic_pfc_counters module in ansible/library/ to get and clear SONiC PFC counters
- Added a new folder test/qos for all QoS related pytest tests
- Added test_pfc_counters.py in test/qos to validate the behavior of PFC counters under PFC pause frames, PFC unpause frames, global pause frames, and global unpause frames. 

#### How did you verify/test it?
Test it in Microsoft internal testbed. 

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
I tested it in a T1 topology but it should support all the topologies.

### CODE REVIEW

File: tests/qos/test_pfc_counters.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    """ Clear PFC counters """
+    dut_ans.sonic_pfc_counters(method = "clear")
+
+    conn_facts = conn_graph_facts['device_conn']	
+    """ Get all the leaf fanout switches """
+    leaf_fanouts = []
+    """ For each interface of DUT"""
+    for intf in conn_facts:
+        peer_device = conn_facts[intf]['peerdevice']
+        if peer_device not in leaf_fanouts:
```
Comment (neethajohn) [inline tests/qos/test_pfc_counters.py]: Did you mean - if peer_device in leaf_fanouts?

File: ansible/library/sonic_pfc_counters.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+					
+	words = line.split()
+	# port_name, counter0, counter1, .... counter7
+	if len(words) != 9:
+	    continue
+		
+	port = words[0]
+	if port not in counters:
+	    counters[port] = dict()
+            counters[port][direction] = [x for x in words[1:]]
```
Comment (neethajohn) [inline ansible/library/sonic_pfc_counters.py]: indentation incorrect

File: ansible/library/sonic_pfc_counters.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	line = line.strip()
+	if 'Port Rx' in line:
+	    direction = rx_direction
+	    continue
+		
+	elif 'Port Tx' in line:
+	    direction = tx_direction
+	    continue
+		
+	if line.startswith('---'):
```
Comment (neethajohn) [inline ansible/library/sonic_pfc_counters.py]: elif

File: tests/qos/test_pfc_counters.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    """
+    @Summary: Map EOS's interface name to Linux's interface name
+    @param eos_intf_name: Interface name in EOS
+    @return: Return the interface name in Linux 
+    """
+    return eos_intf_name.replace('Ethernet', 'et').replace('/', '_')
+
+def test_pfc_pause(ansible_adhoc, testbed, conn_graph_facts):
+    """ @Summary: Run PFC pause frame (pause time quanta > 0) tests """
+    run_pfc(ansible_adhoc, testbed, conn_graph_facts, 65535)
```
Comment (neethajohn) [inline tests/qos/test_pfc_counters.py]: Can you try merging the run_pfc and run_fc? There is a lot of code duplication in both. The only difference is enabling the '-g' to send global pause and counter_facts check. You can do something like this by passing a bool for global pause.

if global pause:
     global_pause = '-g'
     counter_values = ['0'] * PRIO_COUNT
else:
     global_pause = ''
     counter_ values = [str(PKT_COUNT)] * PRIO_COUNT

File: tests/qos/test_pfc_counters.py (hunk 1 above)
Comment (baiwei0427) [reply tests/qos/test_pfc_counters.py]: Here I want to avoid duplicated leaf fanout switches.

Comment (baiwei0427) [reply tests/qos/test_pfc_counters.py]: @neethajohn I want to avoid duplicated leaf fanout switches

File: tests/qos/test_pfc_counters.py (hunk 4 above)
Comment (baiwei0427) [reply tests/qos/test_pfc_counters.py]: @neethajohn I think run_pfc and run_fc are a little bit different. In addition to '-g' option, run_pfc needs to generate PFC frames on all the priorities while run_fc does not.

File: ansible/library/sonic_pfc_counters.py (hunk 3 above)
Comment (baiwei0427) [reply ansible/library/sonic_pfc_counters.py]: @neethajohn  Fixed

File: ansible/library/sonic_pfc_counters.py (hunk 2 above)
Comment (baiwei0427) [reply ansible/library/sonic_pfc_counters.py]: @neethajohn  Fixed

Comment (baiwei0427) [reply ansible/library/sonic_pfc_counters.py]: @neethajohn  Fixed

File: tests/qos/test_pfc_counters.py (hunk 4 above)
Comment (baiwei0427) [reply tests/qos/test_pfc_counters.py]: @neethajohn  Fixed! Merge two functions into run_test

File: tests/qos/test_pfc_counters.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    """ Clear PFC counters """
+    dut_ans.sonic_pfc_counters(method = "clear")
+
+    """ Copy the PFC generator to all the leaf fanout switches """
+    for peer_device in leaf_fanouts:
+        peerdev_ans = AnsibleHost(ansible_adhoc, peer_device)
+        file_src = os.path.join(os.path.dirname(__file__), PFC_GEN_FILE_RELATIVE_PATH)
+        peerdev_ans.copy(src = file_src, dest = PFC_GEN_FILE_DEST, force = True)
+
+def run_test(is_pfc, ansible_adhoc, testbed, conn_graph_facts, leaf_fanouts, pause_time):
```
[the remark this answers is not in the document]
Comment (baiwei0427) [reply tests/qos/test_pfc_counters.py]: @neethajohn  Fixed

**Pull request merged**